Extended OnlineClient information variables

bdeb1337 opened the pull request:
Hello,

First off thank you for this great library, which is working great for a project of mine.

I was having somewhat the same request as [@30](https://github.com/Unity-Technologies/go-ts3/pull/30) of wanting to expand the basic `clientlist` command. I've read the latest serverquery documentation that ships with ts3 server version 3.13.7 and included all parameters regarding this command:

![image](https://github.com/Unity-Technologies/go-ts3/assets/49885109/beb90e46-4e9f-4057-a8bf-b1d3aa91797b)

I haven't included `servergroups` yet, since trying to map it as a `[]int` didn't immediately work. I guess it would have something to do with helpers.go line 74 which states `// TODO(steve): support groups` but I could be wrong. If you are OK with this I could try to give it a go later in a separate pull request? I feel like it might be a little much for one PR otherwise and I could use some guidance in the right direction for that change.

Kind regards

Bob

stevenh reviewed (changes requested):
As not everyone will want all of this information this should really be an extension not a hard coded set.

bdeb1337 commented:
> As not everyone will want all of this information this should really be an extension not a hard coded set.

Thanks for your reply, I agree. I tried implementing this as an optional string parameter for the `ClientList()` function with the ability to request "all". Don't know if there would be a better way of doing this.

stevenh commented on server_cmds.go:
```diff
@@ -351,19 +351,52 @@ func (s *ServerMethods) PrivilegeKeyAdd(ttype, id1, id2 int, options ...CmdArg)
+	Type                           int    `ms:"client_type"`
+	Away                           bool   `ms:"client_away"`
+	AwayMessage                    string `ms:"client_away_message"`
+	FlagTalking                    bool   `ms:"client_flag_talking"`
+	InputMuted                     bool   `ms:"client_input_muted"`
+	OutputMuted                    bool   `ms:"client_output_muted"`
+	InputHardware                  bool   `ms:"client_input_hardware"`
+	OutputHardware                 bool   `ms:"client_output_hardware"`
+	TalkPower                      int    `ms:"client_talk_power"`
+	IsTalker                       bool   `ms:"client_is_talker"`
+	IsPrioritySpeaker              bool   `ms:"client_is_priority_speaker"`
+	IsRecording                    bool   `ms:"client_is_recording"`
+	IsChannelCommander             bool   `ms:"client_is_channel_commander"`
+	UniqueIdentifier               string `ms:"client_unique_identifier"`
+	ChannelGroupID                 int    `ms:"client_channel_group_id"`
+	ChannelGroupInheritedChannelID int    `ms:"client_channel_group_inherited_channel_id"`
+	Version                        string `ms:"client_version"`
+	Platform                       string `ms:"client_platform"`
+	IdleTime                       int    `ms:"client_idle_time"`
+	Created                        int    `ms:"client_created"`
+	LastConnected                  int    `ms:"client_lastconnected"`
+	IconID                         int    `ms:"client_icon_id"`
+	Country                        string `ms:"client_country"`
+	IP                             string `ms:"connection_client_ip"`
+	Badges                         string `ms:"client_badges"`
+	// TODO:
+	// ServerGroups                 []int  `ms:"client_servergroups"`
```
With the additional fields being optional we need a way to ensure users don't mistake a zero value for the real value, thoughts on how to achieve that?

bdeb1337 replied:
Also here thank you for the good point. In the new commit, I changed all optional fields to use pointers, so when they are not returned, it will point to nil instead of a real value. I've also included "Away" and "AwayMessage" to user pointers, since the server would only return these when the `-away` flag is present.

stevenh commented on server_cmds.go:
```diff
@@ -351,19 +351,52 @@ func (s *ServerMethods) PrivilegeKeyAdd(ttype, id1, id2 int, options ...CmdArg)
+	Type                           int    `ms:"client_type"`
+	Away                           bool   `ms:"client_away"`
+	AwayMessage                    string `ms:"client_away_message"`
+	FlagTalking                    bool   `ms:"client_flag_talking"`
+	InputMuted                     bool   `ms:"client_input_muted"`
+	OutputMuted                    bool   `ms:"client_output_muted"`
+	InputHardware                  bool   `ms:"client_input_hardware"`
+	OutputHardware                 bool   `ms:"client_output_hardware"`
+	TalkPower                      int    `ms:"client_talk_power"`
+	IsTalker                       bool   `ms:"client_is_talker"`
+	IsPrioritySpeaker              bool   `ms:"client_is_priority_speaker"`
+	IsRecording                    bool   `ms:"client_is_recording"`
+	IsChannelCommander             bool   `ms:"client_is_channel_commander"`
+	UniqueIdentifier               string `ms:"client_unique_identifier"`
+	ChannelGroupID                 int    `ms:"client_channel_group_id"`
+	ChannelGroupInheritedChannelID int    `ms:"client_channel_group_inherited_channel_id"`
+	Version                        string `ms:"client_version"`
+	Platform                       string `ms:"client_platform"`
+	IdleTime                       int    `ms:"client_idle_time"`
+	Created                        int    `ms:"client_created"`
+	LastConnected                  int    `ms:"client_lastconnected"`
+	IconID                         int    `ms:"client_icon_id"`
+	Country                        string `ms:"client_country"`
+	IP                             string `ms:"connection_client_ip"`
+	Badges                         string `ms:"client_badges"`
+	// TODO:
+	// ServerGroups                 []int  `ms:"client_servergroups"`
```
What's needed to make groups work?

bdeb1337 replied:
The decoding of the server responses wasn't prepared to handle comma-delimited values in responses. Since it is possible to encounter comma's in other fields (a server name or a nickname with a comma for example), I only implemented this for when the key value matches a certain string: I matched on client_servergroups. It is now also possible to receive this information from the server.

bdeb1337 commented on server_cmds.go:
```diff
@@ -351,19 +351,52 @@ func (s *ServerMethods) PrivilegeKeyAdd(ttype, id1, id2 int, options ...CmdArg)
+		} else {
+			clientListParams += " " + param
+		}
+	}
```
Thanks for the suggestion, I agree it would make more sense. I tried adding the flags as variables and edited the code in my new commit.

bdeb1337 commented:
I have added a new commit in which I have tried to address the remarks and questions you made earlier. On the remarks you made above I have added a response with what I wrote this time. Also I have tried editing the *_test.go files to add testing for the standard ClientList and extended Clientlist requests, but I am not that sure if I handled that correctly.

stevenh commented on helpers.go:
```diff
@@ -71,12 +71,23 @@ func DecodeResponse(lines []string, v interface{}) error {
+					// Only support comma seperated lists
+					// by keyname to avoid incorrect decoding.
+					if key == "client_servergroups" {
+						var serverGroups []int
+						for _, s := range strings.Split(v, ",") {
+							if i, err := strconv.Atoi(s); err == nil {
+								serverGroups = append(serverGroups, i)
+							}
+						}
+						input[key] = serverGroups
```
If these are expected to be int's I would return an error if decode fails.

You can also optimise by using the split size.
```suggestion
						var serverGroups []int
						parts := strings.Split(v, ",")
						serverGroups := make([]int, len(parts))
						for i, s := range parts  {
							group, err := strconv.Atoi(s)
							if err != nil {
								return fmt.Errorf("decode server group: %w", err)						
							}
							serverGroups[i] = group
						}
						input[key] = serverGroups
```

bdeb1337 replied:
Thank you for your insights, I have indeed forgotten to return an error. I believe it is correct that these would indeed be expected as int.
I don't really know if I understand the optimalisation correctly: the benefit here is that the array `serverGroups` is an array made just large enough instead of a dynamic length `[]int` array? Or is the benefit coming from the fact that the split is done outside of the range rule? Or are there some other factors at play here? I don't seem to be sure.
I have adopted your code in the new commit ,except for the first line, and it works for me.

stevenh replied:
Correct as `strings.Split` returns the exact size slice we need when we make `serverGroups` to that size, so the runtime can avoid either creating a slice that's too big wasting memory or creating one too small and having the performance penalty of having  to resize and possibly relocate the data.

In addition to that as we know we have the right size, we can assign by index avoiding the overhead of `append`

bdeb1337 replied:
Thank you for explaining this to me.

stevenh commented on mockserver_test.go:
```diff
@@ -256,6 +257,9 @@ func (s *server) handle(conn net.Conn) {
 		l := sc.Text()
 		parts := strings.Split(l, " ")
 		cmd := strings.TrimSpace(parts[0])
+		if cmd == "clientlist" {
```
A little comment here wouldn't go a miss as it's clear from first read why this is needed.

bdeb1337 replied:
The new commit holds a comment where I try to explain why this was necessary for testing with options purposes.

stevenh commented on server_cmds.go:
```diff
@@ -7,6 +7,20 @@ import (
+	ClientInfo    = "-info"
+	ClientIcon    = "-icon"
+	ClientCountry = "-country"
+	ClientIp      = "-ip"
```
```suggestion
	ClientIP      = "-ip"
```

bdeb1337 replied:
Included this in the new commit.

stevenh commented on server_cmds.go:
```diff
@@ -351,19 +365,48 @@ func (s *ServerMethods) PrivilegeKeyAdd(ttype, id1, id2 int, options ...CmdArg)
+func (s *ServerMethods) ClientList(options ...string) ([]*OnlineClient, error) {
 	var clients []*OnlineClient
-	if _, err := s.ExecCmd(NewCmd("clientlist").WithResponse(&clients)); err != nil {
+	var clientListOptions string
+	for _, option := range options {
+		clientListOptions += " " + option
+	}
+	if _, err := s.ExecCmd(NewCmd("clientlist" + clientListOptions).WithResponse(&clients)); err != nil {
```
You should be able to use `WithOptions` here.
```suggestion
	if _, err := s.ExecCmd(NewCmd("clientlist").WithOptions(options).WithResponse(&clients)); err != nil {
```

bdeb1337 replied:
Awesome, didn't realise this was possible. This makes it a lot cleaner, thanks! Included in new commit.

stevenh commented on server_cmds.go:
```diff
@@ -351,19 +365,48 @@ func (s *ServerMethods) PrivilegeKeyAdd(ttype, id1, id2 int, options ...CmdArg)
+	DatabaseID int    `ms:"client_database_id"`
+	Nickname   string `ms:"client_nickname"`
+	Type       int    `ms:"client_type"`
+	// Following variables are optional and can be requested in ClientList() to get extended client information.
```
Would it make sense to split each of these out into separate embedded structs such that they correspond to the command line flags, that way the caller can use minimal dereferencing when consuming?

Thoughts?

bdeb1337 replied:
Sounds like a good idea to me. I tried implementing this in the new commit and personally think it *looks* better. I'm a little too inexperienced to know if this would fulfil the goal of ensuring 'the caller can use minimal dereferencing when consuming'.  I don't think I fully understand what this means yet. Eager to read what your thoughts are on my attempt.

stevenh commented on server_cmds.go:
```diff
@@ -351,19 +365,48 @@ func (s *ServerMethods) PrivilegeKeyAdd(ttype, id1, id2 int, options ...CmdArg)
+	Nickname   string `ms:"client_nickname"`
+	Type       int    `ms:"client_type"`
+	// Following variables are optional and can be requested in ClientList() to get extended client information.
+	Away                           *bool   `ms:"client_away"`
+	AwayMessage                    *string `ms:"client_away_message"`
```
We should specifically not change these two to maintain compatibility.
```suggestion
	Away                           bool   `ms:"client_away"`
	AwayMessage                    string `ms:"client_away_message"`
```

bdeb1337 replied:
Ok. I stopped using a pointer on these two, but I did make a separate embedded struct for these, since they correspond to the `ClientAway` option. As far as I can understand, this wouldn't change the way how they can be addressed from an OnlineClient struct thus maintaining compatibility. Right?

stevenh replied:
Will need to check this as a duplicate could cause problems.

stevenh commented on server_cmds_test.go:
```diff
@@ -306,19 +306,98 @@ func testCmdsServer(t *testing.T, c *Client) {
+				DatabaseID:                     19,
+				Nickname:                       "bdeb1337",
+				Type:                           0,
+				Away:                           nil,
+				AwayMessage:                    nil,
+				FlagTalking:                    nil,
+				InputMuted:                     nil,
+				OutputMuted:                    nil,
+				InputHardware:                  nil,
+				OutputHardware:                 nil,
+				TalkPower:                      nil,
+				IsTalker:                       nil,
+				IsPrioritySpeaker:              nil,
+				IsRecording:                    nil,
+				IsChannelCommander:             nil,
+				UniqueIdentifier:               nil,
+				ChannelGroupID:                 nil,
+				ChannelGroupInheritedChannelID: nil,
+				Version:                        nil,
+				Platform:                       nil,
+				IdleTime:                       nil,
+				Created:                        nil,
+				LastConnected:                  nil,
+				IconID:                         nil,
+				Country:                        nil,
+				IP:                             nil,
+				Badges:                         nil,
+				ServerGroups:                   nil,
```
Since nil is the empty value no need to define.
```suggestion
```

bdeb1337 replied:
Thank you. Understood & edited in the new commit.

stevenh commented on server_cmds_test.go:
```diff
@@ -306,19 +306,98 @@ func testCmdsServer(t *testing.T, c *Client) {
+				DatabaseID:                     19,
+				Nickname:                       "bdeb1337",
+				Type:                           0,
+				Away:                           nil,
+				AwayMessage:                    nil,
```
Did the default change here, not sure why?

bdeb1337 replied:
I tend to also think in your direction that it changed over time in some version. I might try to see if I can pinpoint that some way by running some queries on older versions perhaps.

stevenh reviewed (changes requested):
Thanks for working on this, making good progress.

bdeb1337 commented:
I have added a new commit trying to implement your recommendations and remarks. Kind regards.

stevenh commented on mockserver_test.go:
```diff
@@ -256,6 +257,11 @@ func (s *server) handle(conn net.Conn) {
 		l := sc.Text()
 		parts := strings.Split(l, " ")
 		cmd := strings.TrimSpace(parts[0])
+		// to support server commands with specific optional parameters,
```
Small tweak, sentences are good
```suggestion
		// Support server commands with specific optional parameters,
```

bdeb1337 replied:
Should be edited in the new commit.

stevenh commented on server_cmds.go:
```diff
@@ -351,19 +365,66 @@ func (s *ServerMethods) PrivilegeKeyAdd(ttype, id1, id2 int, options ...CmdArg)
+	// Following variables are optional and can be requested in ClientList() to get extended client information.
+	// note: Away and AwayMessage are currently optional but not using pointers for compatibility with older versions of teamspeakserver.
+	// This means they will always be returned, but their value will only be representative when the ClientAway option is specified.
+	OnlineClientAway   `ms:",squash"` // ts3.ClientAway
+	UniqueIdentifier   *string        `ms:"client_unique_identifier"` // ts3.ClientUid
+	OnlineClientVoice  `ms:",squash"` // ts3.ClientVoice
+	OnlineClientTimes  `ms:",squash"` // ts3.ClientTimes
+	OnlineClientGroups `ms:",squash"` // ts3.ClientGroups
+	OnlineClientInfo   `ms:",squash"` // ts3.ClientInfo
```
We can make the embeds pointers to avoid duplicate decoding checks and pointer dereference.

While not ideal as we're at v1.xx we should fully revert the away changes to ensure full compatibility.

I'm assuming with `OnlineClientAway` won't be populated without `ClientAway` option? If so we should comment to make that clear.

Also as we're in the package we shouldn't include ts3 in the comment.

Note: suggestion may need reformatting
```suggestion
       	Away        bool   `ms:"client_away"` // Only populated if ClientAway or ClientListFull is passed to ClientList.
	AwayMessage string `ms:"client_away_message"` // Only populated if ClientAway or ClientListFull is passed to ClientList.
	UniqueIdentifier   *string        `ms:"client_unique_identifier"` // Only populated if ClientUid or ClientListFull is passed to ClientList.
	*OnlineClientVoice  `ms:",squash"` // Only populated if ClientVoice or ClientListFull is passed to ClientList. 
	*OnlineClientTimes  `ms:",squash"` // Only populated if ClientTimes or ClientListFull is passed to ClientList. 
	*OnlineClientGroups `ms:",squash"` // Only populated if ClientGroups or ClientListFull is passed to ClientList. 
	*OnlineClientInfo   `ms:",squash"` // Only populated if ClientInfo or ClientListFull is passed to ClientList.
```

bdeb1337 replied:
> We can make the embeds pointers to avoid duplicate decoding checks and pointer dereference.

I have tried implementing this in the newest commit.

> While not ideal as we're at v1.xx we should fully revert the away changes to ensure full compatibility.

Ok, I understand. I should have now reverted this to how it used to be originally.

> I'm assuming with OnlineClientAway won't be populated without ClientAway option? If so we should comment to make 
that clear.

Agreed, added in comment in new commit.

> Also as we're in the package we shouldn't include ts3 in the comment.

I have taken your example and added the new comments in the new commit.

stevenh commented on server_cmds_test.go:
```diff
@@ -306,19 +306,87 @@ func testCmdsServer(t *testing.T, c *Client) {
+			return
+		}
+
+		// helper functions to return pointers
+		boolptr := func(b bool) *bool {
+			return &b
+		}
+		stringptr := func(s string) *string {
+			return &s
+		}
+		intptr := func(i int) *int {
+			return &i
+		}
```
With the change to pointer embeds these might not be needed.

bdeb1337 replied:
I think they are still needed, except for the boolptr function, I have made variables for this. Unless there would be another way that you think is better?

stevenh commented on server_cmds.go:
```diff
@@ -351,19 +365,66 @@ func (s *ServerMethods) PrivilegeKeyAdd(ttype, id1, id2 int, options ...CmdArg)
+	OnlineClientTimes  `ms:",squash"` // ts3.ClientTimes
+	OnlineClientGroups `ms:",squash"` // ts3.ClientGroups
+	OnlineClientInfo   `ms:",squash"` // ts3.ClientInfo
+	IconID             *int           `ms:"client_icon_id"`       // ts3.ClientIcon
+	Country            *string        `ms:"client_country"`       // ts3.ClientCountry
+	IP                 *string        `ms:"connection_client_ip"` // ts3.ClientIP
+	Badges             *string        `ms:"client_badges"`        // ts3.ClientBadges
```
Similar updates to comments as above.

bdeb1337 replied:
Also here I followed your example and changed the comments.

bdeb1337 commented:
In the new commit, an embedded struct pointer that holds all optional struct pointers and values for options of clientlist has been added. The `mapstructure` library didn't support the combination of reflection and nil values to determine the type when squashing embedded pointers from map to struct. But it does work with regular structs: (https://github.com/mitchellh/mapstructure/blob/bf980b35cac4dfd34e05254ee5aba086504c3f96/mapstructure_test.go#L729C17-L729C54).
I have added a workaround by injecting the correct type while decoding, and nil'ling them after if needed.

bdeb1337 commented:
Hi @stevenh , I hope you are doing well. 
Time has gone by fast since September and I apologize for not bumping this earlier. I was wondering if you have had a chance to look at my latest commit and share your thoughts about the implementation. 
I appreciate your feedback and suggestions because I already learned a lot from you. 
Thank you for your time and attention.
Kind regards!

stevenh commented:
Thanks for the bump @bdeb1337 will try to look over the next few days, but as I no longer have merge access, also poking @lwaddicor

lwaddicor commented:
Hey!

Just had a quick look and ran the change against a local teamspeak server I fired up in docker, and it looks as populated as I'd expect.

Validation
```
package main

import (
	"fmt"
	"log"

	"github.com/kr/pretty"
	"github.com/multiplay/go-ts3"
)

func main() {
	c, err := ts3.NewClient("localhost:10011")
	if err != nil {
		log.Fatal(err)
	}
	defer c.Close()

	if err := c.Login("serveradmin", "M2EhMOoP"); err != nil {
		log.Fatal(err)
	}

	v, err := c.Version()
	if err != nil {
		log.Fatal(fmt.Errorf("fail to version list: %w", err))
	}
	fmt.Println("Server is running:")
	pretty.Print(v)

	if err := c.Server.Use(1); err != nil {
		log.Fatal(err)
	}

	fmt.Println("Client fist full:")
	cl, err := c.Server.ClientList(ts3.ClientListFull)
	if err != nil {
		log.Fatal(fmt.Errorf("fail to client list: %w", err))
	}
	pretty.Print(cl)
}
```
Output:
```
Server is running:
&ts3.Version{Version:"3.13.7", Platform:"Linux", Build:1655727713}
Client fist full:
[]*ts3.OnlineClient{
    &ts3.OnlineClient{
        ID:              13,
        ChannelID:       1,
        DatabaseID:      1,
        Nickname:        "serveradmin",
        Type:            1,
        Away:            false,
        AwayMessage:     "",
        OnlineClientExt: &ts3.OnlineClientExt{
            UniqueIdentifier:  &"serveradmin",
            OnlineClientVoice: &ts3.OnlineClientVoice{
                FlagTalking:        &bool(false),
                InputMuted:         &bool(false),
                OutputMuted:        &bool(false),
                InputHardware:      &bool(false),
                OutputHardware:     &bool(false),
                TalkPower:          &int(0),
                IsTalker:           &bool(false),
                IsPrioritySpeaker:  &bool(false),
                IsRecording:        &bool(false),
                IsChannelCommander: &bool(false),
            },
            OnlineClientTimes: &ts3.OnlineClientTimes{
                IdleTime:      &int(6),
                Created:       &int(0),
                LastConnected: &int(0),
            },
            OnlineClientGroups: &ts3.OnlineClientGroups{
                ChannelGroupID:                 &int(8),
                ChannelGroupInheritedChannelID: &int(1),
                ServerGroups:                   &[]int{2},
            },
            OnlineClientInfo: &ts3.OnlineClientInfo{
                Version:  &"ServerQuery",
                Platform: &"ServerQuery",
            },
            Country: &"",
            IP:      &"172.17.0.1",
            Badges:  &"",
            IconID:  &int(0),
        },
    },
    &ts3.OnlineClient{
        ID:              6,
        ChannelID:       1,
        DatabaseID:      3,
        Nickname:        "Lewis",
        Type:            0,
        Away:            true,
        AwayMessage:     "",
        OnlineClientExt: &ts3.OnlineClientExt{
            UniqueIdentifier:  &"lnxjPE/99+DR/wFXWawTtpeMBg4=",
            OnlineClientVoice: &ts3.OnlineClientVoice{
                FlagTalking:        &bool(false),
                InputMuted:         &bool(false),
                OutputMuted:        &bool(false),
                InputHardware:      &bool(true),
                OutputHardware:     &bool(true),
                TalkPower:          &int(75),
                IsTalker:           &bool(false),
                IsPrioritySpeaker:  &bool(false),
                IsRecording:        &bool(false),
                IsChannelCommander: &bool(true),
            },
            OnlineClientTimes: &ts3.OnlineClientTimes{
                IdleTime:      &int(312470),
                Created:       &int(1701366169),
                LastConnected: &int(1701366169),
            },
            OnlineClientGroups: &ts3.OnlineClientGroups{
                ChannelGroupID:                 &int(8),
                ChannelGroupInheritedChannelID: &int(1),
                ServerGroups:                   &[]int{6},
            },
            OnlineClientInfo: &ts3.OnlineClientInfo{
                Version:  &"3.6.1 [Build: 1690193193]",
                Platform: &"OS X",
            },
            Country: &"",
            IP:      &"172.17.0.1",
            Badges:  &"",
            IconID:  &int(0),
        },
    },
}
```

bdeb1337 commented on server_cmds.go:
```diff
@@ -351,19 +365,65 @@ func (s *ServerMethods) PrivilegeKeyAdd(ttype, id1, id2 int, options ...CmdArg)
+	*OnlineClientExt `ms:",squash"` // Only populated if any of the options is passed to ClientList.
+}
+
+type OnlineClientExt struct {
```
Thanks for noticing, I think I've added all missing comments in the latest commit.

bdeb1337 commented on server_cmds.go:
```diff
@@ -7,6 +7,20 @@ import (
+	ClientIcon    = "-icon"
+	ClientCountry = "-country"
+	ClientIP      = "-ip"
+	ClientBadges  = "-badges"
```
I didn't know that these had influence on the docs, thanks for showing me this. I was wondering what linter you would recommend? It seems like I had `staticcheck`, and tried using `golint` for the first time and indeed it also set it off for me :). My commit seems to have another indent as your suggestion, is that also something worth looking into?

bdeb1337 commented:
> Hey!
> 
> Just had a quick look and ran the change against a local teamspeak server I fired up in docker, and it looks as populated as I'd expect.
> 
> Just a couple of minor things around the comments to please the linters and make it easier for users in their IDEs. Other than that it looks very nice!

Thanks a bunch for checking it out so quickly! I've added your latest suggestions in a new commit.

bdeb1337 commented:
Hi @lwaddicor and @stevenh 

My best wishes for the starting of the new year :fireworks: .  I would kindly like to bump this PR again as I am looking forward to what you guys think and what I should try to improve or change.

Thank you!
Kind regards

lwaddicor reviewed (approved):
Sorry for the delay. Looks good to me

bdeb1337 commented:
No problem, thank you @lwaddicor for checking and merging. And also @stevenh thanks again for the guidance in this PR, appreciate that. 
Kind regards, 
Bob.
